File: platforms/ios/Packages/FunputKit/Sources/KeyboardLayout/Alternates/VietnameseKeyAlternates.swift
public enum VietnameseKeyAlternates {
    public static func values(for character: Character) -> [KeyAlternate] {
        guard let values = catalog[character.lowercased().first ?? character] else { return [] }
        return values.map { KeyAlternate(text: $0) }
    }

    private static let catalog: [Character: [String]] = [
        "a": groups("aáàảãạ", "ăắằẳẵặ", "âấầẩẫậ"),
        "e": groups("eéèẻẽẹ", "êếềểễệ"),
        "i": groups("iíìỉĩị"),
        "o": groups("oóòỏõọ", "ôốồổỗộ", "ơớờởỡợ"),
        "u": groups("uúùủũụ", "ưứừửữự"),
        "y": groups("yýỳỷỹỵ"),
        "d": groups("dđ"),
    ]

    private static func groups(_ values: String...) -> [String] {
        values.flatMap { $0.map(String.init) }
    }
}

File: platforms/android/keyboard-renderer/src/main/java/app/funput/funput/keyboard/popover/model/VietnameseKeyAlternates.kt
package app.funput.funput.keyboard.popover.model

internal object VietnameseKeyAlternates {
    fun valuesFor(character: Char): List<KeyAlternate> =
        Catalog[character.lowercaseChar()].orEmpty().map(::KeyAlternate)

    private val Catalog = mapOf(
        'a' to groups("aáàảãạ", "ăắằẳẵặ", "âấầẩẫậ"),
        'e' to groups("eéèẻẽẹ", "êếềểễệ"),
        'i' to groups("iíìỉĩị"),
        'o' to groups("oóòỏõọ", "ôốồổỗộ", "ơớờởỡợ"),
        'u' to groups("uúùủũụ", "ưứừửữự"),
        'y' to groups("yýỳỷỹỵ"),
        'd' to groups("dđ"),
    )

    private fun groups(vararg values: String): List<String> =
        values.flatMap { group -> group.map(Char::toString) }
}
